File: api/src/core/security/JWT.go
package security

import (
	"errors"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

const TokenTTL = 24 * time.Hour

var ErrInvalidToken = errors.New("token invalido o expirado")

type Claims struct {
	UserID string `json:"user_id"`
	Role   string `json:"role"`
	jwt.RegisteredClaims
}

func GenerateToken(userID string, role string, secret string) (string, error) {
	claims := Claims{
		UserID: userID,
		Role:   role,
		RegisteredClaims: jwt.RegisteredClaims{
			IssuedAt:  jwt.NewNumericDate(time.Now()),
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(TokenTTL)),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(secret))
}

func ParseToken(tokenString string, secret string) (*Claims, error) {
	claims := &Claims{}

	token, err := jwt.ParseWithClaims(tokenString, claims, func(t *jwt.Token) (interface{}, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, ErrInvalidToken
		}
		return []byte(secret), nil
	})
	if err != nil || !token.Valid {
		return nil, ErrInvalidToken
	}

	return claims, nil
}

File: realtime/src/core/security/JWT.ts
import jwt from "jsonwebtoken";

export interface Claims {
  userId: string;
  role: string;
}

export function verifyToken(token: string, secret: string): Claims {
  const decoded = jwt.verify(token, secret) as jwt.JwtPayload;

  if (typeof decoded.user_id !== "string" || decoded.user_id === "") {
    throw new Error("token invalido: falta user_id");
  }

  return {
    userId: decoded.user_id,
    role: typeof decoded.role === "string" ? decoded.role : "",
  };
}
